File: emm_g1/controller.py
"""Environment-aware motion-matching controller (Ponton et al. 2025), G1 port.

One controller drives one kinematic G1. Each step it predicts a desired
trajectory from the input (spring smoothing), builds the query feature vector,
runs the environment-aware search (:mod:`emm_g1.search`) against the database,
inertializes into the matched frame, and integrates the root from the matched
clip's root velocity. The obstacle **penalization** with **height gating** is what
makes a low bar trigger a jump with no manual command. Direction-weight
**evasion** and environment **anticipation** reproduce the paper's adaptive
behaviour.

``step(left_stick, right_stick)`` returns a world-space ``qpos (36,)`` (wxyz
quat), the same layout ``mm_g1`` produces, so the viewer / web port are uniform.
"""

import numpy as np

from . import quat
from . import g1_model as g1
from . import springs as sp
from . import search as S
from . import features as feat   # noqa: F401  (kept for parity / external use)

# Static feature layout (27): [trajPos 0:6, trajDir 6:12, pose 12:27].
_TRAJPOS = slice(0, 6)
_TRAJDIR = slice(6, 12)
_POSE = slice(12, 27)


def default_weights():
    w = np.ones(27, np.float32)
    w[_TRAJPOS] = 1.0
    w[_TRAJDIR] = 1.5      # facing -- modulated down by evasion near obstacles
    w[12:15] = 1.0         # left foot position
    w[15:18] = 0.25        # left foot velocity
    w[18:21] = 1.0         # right foot position
    w[21:24] = 0.25        # right foot velocity
    w[24:27] = 0.25        # hips velocity
    return w


class EMMController:
    def __init__(self, db, env=None, weights=None, max_speed=1.7,
                 search_time=0.12, inert_halflife=0.10, vel_halflife=0.25,
                 rot_halflife=0.25, penalty_weight=18.0, evasion=None,
                 anticipation=2.0, start_frame=None, trigger_dist=0.85):
        self.db = S.prepare(db)
        self.env = env
        self.height_mode = bool(np.asarray(db['height_mode']))
        self.w0 = (weights if weights is not None else default_weights()).astype(np.float32)
        self.w = self.w0.copy()
        self.start_dir_w = float(self.w0[_TRAJDIR][0])
        self.MAX_SPEED = max_speed
        self.SEARCH_TIME = search_time
        self.INERT = inert_halflife
        self.VELH = vel_halflife
        self.ROTH = rot_halflife
        self.penalty_weight = penalty_weight
        # Evasion default: 0.54 for the height (jump/crouch) scenario, else 0.01
        # (reference Search .asset values).
        self.evasion = (0.54 if self.height_mode else 0.01) if evasion is None else evasion
        self.anticipation = anticipation

        # --- Jump skill bucket: a separate bucket from locomotion (mm_g1 style).
        # The search runs over locomotion frames only (loco_mask); the jump is
        # entered through its run-up when an obstacle sits in the take-off window
        # ahead -- the obstacle replaces mm_g1's J key. ---
        from . import jumps as _J
        self.jump_enter, self.jump_land_of, self.jump_end_of, self.jump_apex_of = _J.jump_index(db)
        self.loco_mask = ~np.asarray(db['clip_is_jump'], bool)[db['clip_id']]
        self.trigger_dist = float(trigger_dist)
        self.jump_locked = 0
        self._prev_fd = np.inf

        self.starts, self.stops = db['starts'], db['stops']
        self.Xn, self.Xmean, self.Xstd = db['Xn'], db['Xmean'], db['Xstd']
        self.dof, self.dofVel = db['dof'], db['dofVel']
        self.rootPosDB, self.rootQuatDB = db['rootPos'], db['rootQuat']
        self.thetaDB, self.rootVelDB, self.yawRateDB = db['theta'], db['rootVel'], db['yawRate']
        self.Ttimes = g1.HORIZONS / g1.FPS
        self._start_frame = start_frame
        self.spawn_xy = None        # if set, reset() places the root here
        self.spawn_yaw = 0.0
        self.reset(start_frame)

    def set_spawn(self, x, y, yaw=0.0):
        """Pin the reset (Space) pose to a world xy + heading (e.g. the lane start)."""
        self.spawn_xy = (float(x), float(y))
        self.spawn_yaw = float(yaw)
        self.reset()

    # -- state --
    def reset(self, start_frame=None):
        if start_frame is None:
            start_frame = self._start_frame
        f0 = int(self.starts[0] + 30) if start_frame is None else int(start_frame)
        self.animClip = int(np.searchsorted(self.stops, f0, side='right'))
        self.animFrame = f0
        self.rootPos = self.rootPosDB[f0].astype(float).copy()
        self.rootYaw = float(self.thetaDB[f0])
        if getattr(self, "spawn_xy", None) is not None:
            self.rootPos[0:2] = self.spawn_xy
            self.rootYaw = self.spawn_yaw
        self.rootRot = g1.yaw_quat(self.rootYaw)
        self.rootVel = np.zeros(3); self.rootAcc = np.zeros(3); self.rootAng = np.zeros(3)
        self.desiredDir = quat.mul_vec(self.rootRot, g1.FORWARD)
        self.offDof = np.zeros(g1.NDOF); self.offDofVel = np.zeros(g1.NDOF)
        self.offH = 0.0; self.offHVel = 0.0
        self.searchTimer = 0.0
        self.jump_locked = 0
        self._prev_fd = np.inf
        self.target_speed = 0.0
        self.w = self.w0.copy()
        self.Tpos = np.tile(self.rootPos, (len(g1.HORIZONS), 1))
        self.Tdir = np.tile(self.desiredDir, (len(g1.HORIZONS), 1))

    @property
    def cur(self):
        return self.animFrame

    @property
    def jumping(self):
        # True only while riding the jump skill (run-up -> flight -> landing),
        # not merely when the playhead happens to sit in a jump clip.
        return self.jump_locked > 0

    def clip_name(self, f=None):
        return str(self.db['clip_names'][self._clip_of(self.animFrame if f is None else f)])

    def _clip_of(self, frame):
        return int(np.searchsorted(self.stops, frame, side='right'))

    def _query(self, Tpos, Tdir, qh_ctrl):
        pose = self.Xn[self.animFrame, _POSE] * self.Xstd[_POSE] + self.Xmean[_POSE]
        trajPos = quat.inv_mul_vec(qh_ctrl, Tpos - self.rootPos)[:, 0:2].ravel()
        trajDir = quat.inv_mul_vec(qh_ctrl, Tdir)[:, 0:2].ravel()
        q = np.concatenate([trajPos, trajDir, pose])
        return ((q - self.Xmean) / self.Xstd).astype(np.float32)

    # -- jump bucket (obstacle-triggered, mm_g1-style run-up + lock) --
    def _forward_obstacle_dist(self):
        """Nearest forward distance (m) to an obstacle whose height band a standing
        body overlaps, along the current facing; ``inf`` if none ahead. An obstacle
        in the take-off window arms the auto-jump (replaces mm_g1's J key)."""
        if self.env is None:
            return np.inf
        fdir = np.array([np.cos(self.rootYaw), np.sin(self.rootYaw)])
        perp = np.array([-fdir[1], fdir[0]])
        best = np.inf
        for o in self.env.obstacles:
            if o.height[0] > 1.3 or o.height[1] < 0.0:     # not in standing-body band
                continue
            rel = o.center - self.rootPos[0:2]
            fwd = float(rel @ fdir)
            if fwd <= 0.0:                                  # behind us
                continue
            reach = (float(max(o.ext)) if o.is_ellipse else o.radius) + 0.5
            if abs(float(rel @ perp)) > reach:             # off to the side
                continue
            best = min(best, fwd)
        return best

    def _best_jump_entry(self):
        """Run-up frame whose features best continue the current pose, so the jump
        is entered from the run-up (smooth take-off), not mid-stride."""
        if len(self.jump_enter) == 0:
            return None
        d = np.linalg.norm(self.Xn[self.jump_enter] - self.Xn[self.animFrame], axis=1)
        return int(self.jump_enter[int(d.argmin())])

    def _switch_to(self, b):
        """Inertialize the pose discontinuity from the current frame into frame ``b``
        and move the playhead there (same offsets the env search uses)."""
        a = self.animFrame
        self.offDof = (self.offDof + self.dof[a]) - self.dof[b]
        self.offDofVel = (self.offDofVel + self.dofVel[a]) - self.dofVel[b]
        self.offH = (self.offH + self.rootPosDB[a, 2]) - self.rootPosDB[b, 2]
        self.animFrame = b
        self.animClip = self._clip_of(b)

    # -- one real-time frame --
    def step(self, left_stick, right_stick):
        db, starts, stops = self.db, self.starts, self.stops
        desiredVel = self.MAX_SPEED * np.asarray(left_stick, float)
        self.target_speed = float(np.linalg.norm(desiredVel))
        if np.linalg.norm(right_stick) > 0.01:
            self.desiredDir = np.asarray(right_stick, float) / np.linalg.norm(right_stick)
        elif np.linalg.norm(left_stick) > 0.01:
            self.desiredDir = np.asarray(left_stick, float) / np.linalg.norm(left_stick)
        desiredRot = g1.yaw_quat(np.arctan2(self.desiredDir[1], self.desiredDir[0]))

        # Desired trajectory via springs.
        dt_col = self.Ttimes[:, None]
        Tpos, _, _ = sp.trajectory_spring_position(
            self.rootPos, self.rootVel, self.rootAcc, desiredVel, self.VELH, dt_col)
        Trot, _ = sp.trajectory_spring_rotation(
            self.rootRot, self.rootAng, desiredRot, self.ROTH, dt_col)
        Tdir = quat.mul_vec(Trot, g1.FORWARD)
        self.Tpos, self.Tdir = Tpos, Tdir

        # Nearby obstacles, per trajectory tap.
        circ = [np.zeros((0, 5))] * 3
        ell = [np.zeros((0, 8))] * 3
        if self.env is not None:
            circ, ell = self.env.get_nearby(self.rootPos[0:2], self.rootYaw,
                                            self.height_mode, tap_world=Tpos[:, 0:2])

        # Jump bucket trigger: arm a jump when the wall ahead is exactly the distance
        # the best-matching run-up clip travels from entry to its flight apex, so the
        # apex (whole body above the wall band) lands over the wall. Entered from the
        # run-up and ridden through flight + landing (search locked) -- like mm_g1's J
        # key, but auto-triggered and auto-timed by the obstacle.
        if self.jump_locked == 0 and self.env is not None:
            fd = self._forward_obstacle_dist()
            # Fire the instant the wall crosses the take-off distance, so every wall
            # is launched from the same spot and the flight's whole-body-clear window
            # lands over it. Enter from the run-up that best continues the pose.
            if self._prev_fd > self.trigger_dist >= fd:
                entry = self._best_jump_entry()
                if entry is not None:
                    self._switch_to(entry)
                    self.jump_locked = max(1, self.jump_end_of[entry] - entry)
                    self.searchTimer = self.SEARCH_TIME
            self._prev_fd = fd

        # Search (rate limited, LOCOMOTION bucket only): env-aware nearest pose.
        if self.jump_locked == 0 and self.searchTimer <= 0.0:
            qh_ctrl = g1.yaw_quat(self.rootYaw)
            Xq = self._query(Tpos, Tdir, qh_ctrl)
            pw = self.penalty_weight * self.anticipation * max(0.5, self.target_speed)
            best = S.search_env(db, Xq, self.w, pw, circ, ell,
                                self.env.threshold if self.env else 0.6,
                                self.height_mode, cand_mask=self.loco_mask)
            if best != self.animFrame:
                self.offDof = (self.offDof + self.dof[self.animFrame]) - self.dof[best]
                self.offDofVel = (self.offDofVel + self.dofVel[self.animFrame]) - self.dofVel[best]
                self.offH = (self.offH + self.rootPosDB[self.animFrame, 2]) - self.rootPosDB[best, 2]
                self.animFrame = best
                self.animClip = self._clip_of(best)
            self.searchTimer = self.SEARCH_TIME

        # Evasion: shrink the facing weight when an obstacle is close.
        self._update_evasion(circ, ell)

        # Advance playhead.
        self.animFrame = int(np.clip(self.animFrame + 1,
                                     starts[self.animClip], stops[self.animClip] - 1))
        self.searchTimer -= g1.DT
        if self.jump_locked > 0:
            self.jump_locked -= 1
            if self.jump_locked == 0:
                self.searchTimer = 0.0           # search out of the jump at once
        elif self.animFrame >= stops[self.animClip] - 2:
            self.searchTimer = 0.0

        # Integrate controller root from the matched clip's root velocity.
        _, _, self.rootAcc = sp.trajectory_spring_position(
            self.rootPos, self.rootVel, self.rootAcc, desiredVel, self.ROTH, g1.DT)
        qh_clip = g1.yaw_quat(self.thetaDB[self.animFrame])
        clipVelLocal = quat.inv_mul_vec(qh_clip, self.rootVelDB[self.animFrame])
        self.rootVel = quat.mul_vec(self.rootRot, clipVelLocal)
        self.rootAng = np.array([0.0, 0.0, self.yawRateDB[self.animFrame]])
        self.rootPos = self.rootPos + self.rootVel * g1.DT
        self.rootYaw = self.rootYaw + self.yawRateDB[self.animFrame] * g1.DT
        self.rootRot = g1.yaw_quat(self.rootYaw)

        # Inertialized pose.
        self.offDof, self.offDofVel = sp.decay_spring_damper(
            self.offDof, self.offDofVel, self.INERT, g1.DT)
        self.offH, self.offHVel = sp.decay_spring_damper(
            self.offH, self.offHVel, self.INERT, g1.DT)
        f = self.animFrame
        dofOut = self.dof[f] + self.offDof
        tilt = quat.mul(quat.inv(qh_clip), self.rootQuatDB[f])
        rootQuatOut = quat.mul(self.rootRot, tilt)

        qpos = np.empty(36)
        qpos[0:2] = self.rootPos[0:2]
        qpos[2] = self.rootPosDB[f, 2] + self.offH
        qpos[3:7] = rootQuatOut
        qpos[7:] = dofOut
        return qpos

    def _update_evasion(self, circ, ell):
        """Lerp the facing-weight down near the closest obstacle (reference
        ``OnSearchCompleted``)."""
        from . import ellipse_geom as eg
        thr = self.env.threshold if self.env else 0.4
        closest = np.inf
        f = self.animFrame
        for k in range(3):
            center = self.db['posTaps'][f, k]
            ax = self.db['ellAxis'][f, k]
            ext = np.array([self.db['ellExtP'][f, k], self.db['ellExtS'][f, k]])
            if len(circ[k]):
                d, _ = eg.point_to_ellipse(center, ax, ext, circ[k][:, 0:2])
                d = np.maximum(d - circ[k][:, 2], eg.MAX_INSIDE)
                dmin = float(d.min())
                if dmin < thr:
                    closest = min(closest, dmin)
            for j in range(len(ell[k])):
                d, _, _ = eg.ellipse_to_ellipse(center, ax, ext, ell[k][j, 0:2],
                                                ell[k][j, 2:4], ell[k][j, 4:6])
                if d < thr:
                    closest = min(closest, d)
        target = self.start_dir_w
        if np.isfinite(closest) and closest < thr:
            factor = np.log10(max(closest / thr, 1e-3)) + 1.0
            target = self.start_dir_w * max(self.evasion, factor)
            rate = np.clip(g1.DT * 10.0, 0.0, 1.0)
        else:
            rate = np.clip(g1.DT * 100.0, 0.0, 1.0)
        cur = self.w[_TRAJDIR][0]
        self.w[_TRAJDIR] = cur + (target - cur) * rate

File: docs/js/emm/controller.js
// Environment-aware motion-matching controller in the browser.
// A 1:1 port of emm_g1/controller.py (EMMController): spring-predicted trajectory
// -> query vector -> env-aware search (obstacle penalization + height gating) ->
// inertialized pose. step(leftStick, rightStick) returns a world-space qpos(36).
// A low wall makes the G1 hop over it with no trigger.

import { quat, v3 } from '../quat.js';
import { Environment } from './obstacles.js';
import { searchEnv } from './search.js';
import { jumpIndex } from './jumps.js';
import * as eg from './ellipse_geom.js';
import {
  decaySpringDamper, decaySpringDamperScalar,
  trajectorySpringPosition, trajectorySpringRotation,
} from './springs.js';

const FORWARD = [1, 0, 0];
const yaw = (t) => quat.yaw(t);

// typed-array views into emm.bin per the emm.json header.
export function loadDB(meta, buf) {
  const A = {};
  for (const [name, h] of Object.entries(meta.arrays)) {
    const TA = h.dtype === 'int32' ? Int32Array : Float32Array;
    const count = h.shape.reduce((a, b) => a * b, 1);
    A[name] = count ? new TA(buf, h.offset, count) : new TA(0);
  }
  A.T = meta.n_frames;
  return A;
}

const clamp = (x, lo, hi) => Math.max(lo, Math.min(hi, x));

export class EMMController {
  constructor(meta, A) {
    this.A = A; this.db = A;
    this.fps = meta.fps; this.DT = 1 / meta.fps;
    this.horizons = meta.horizons;
    this.Ttimes = meta.horizons.map((h) => h / meta.fps);
    this.MAX_SPEED = meta.max_speed; this.WALK_SCALE = meta.walk_scale;
    this.SEARCH_TIME = meta.search_time;
    this.INERT = meta.inert_halflife; this.VELH = meta.vel_halflife; this.ROTH = meta.rot_halflife;
    this.penaltyWeight = meta.penalty_weight; this.evasion = meta.evasion;
    this.anticipation = meta.anticipation; this.threshold = meta.threshold;
    this.heightMode = !!meta.height_mode;
    this.clipNames = meta.clip_names;

    this.Xn = A.Xn; this.Xmean = A.Xmean; this.Xstd = A.Xstd;
    this.dof = A.dof; this.dofVel = A.dofVel;
    this.rootPosDB = A.rootPos; this.rootQuatDB = A.rootQuat;
    this.thetaDB = A.theta; this.rootVelDB = A.rootVel; this.yawRateDB = A.yawRate;
    this.starts = A.starts; this.stops = A.stops;
    this.clipIsJump = A.clip_is_jump;

    this.w0 = Array.from(A.weights);          // base 27 weights
    this.startDirW = this.w0[6];
    this.env = new Environment(meta.obstacles, meta.threshold);

    // --- Jump skill bucket (separate from locomotion, like mm_g1): the search runs
    // over locomotion frames only (locoMask); the jump is entered through its run-up
    // when an obstacle crosses triggerDist ahead -- the obstacle replaces the J key. ---
    const ji = jumpIndex(A, this.clipNames);
    this.jumpEnter = ji.enter; this.landOf = ji.landOf; this.endOf = ji.endOf;
    this.locoMask = new Uint8Array(A.T);
    for (let f = 0; f < A.T; f++) this.locoMask[f] = A.clip_is_jump[A.clip_id[f]] === 1 ? 0 : 1;
    this.triggerDist = meta.trigger_dist !== undefined ? meta.trigger_dist : 0.8;

    this.reset();
  }

  _row(arr, d, f) { const b = f * d, r = new Array(d); for (let i = 0; i < d; i++) r[i] = arr[b + i]; return r; }
  _clipOf(f) { let r = 0; for (let i = 0; i < this.starts.length; i++) { if (this.stops[i] <= f) r = i + 1; else break; } return r; }

  reset(startFrame = null) {
    const f0 = startFrame === null ? this.starts[0] + 30 : startFrame;
    this.animClip = this._clipOf(f0);
    this.animFrame = f0;
    this.rootPos = this._row(this.rootPosDB, 3, f0);
    this.rootYaw = this.thetaDB[f0];
    this.rootRot = yaw(this.rootYaw);
    this.rootVel = [0, 0, 0]; this.rootAcc = [0, 0, 0]; this.rootAng = [0, 0, 0];
    this.desiredDir = quat.mulVec(this.rootRot, FORWARD);
    this.offDof = new Array(29).fill(0); this.offDofVel = new Array(29).fill(0);
    this.offH = 0; this.offHVel = 0;
    this.searchTimer = 0; this.targetSpeed = 0;
    this.jumpLocked = 0; this.prevFd = Infinity;
    this.w = this.w0.slice();
    this.Tpos = this.Ttimes.map(() => this.rootPos.slice());
    this.Tdir = this.Ttimes.map(() => this.desiredDir.slice());
  }

  // Spawn helper (matches run_emm.py): place root at xy, facing +x lane.
  spawn(x, y, yawRad = 0) {
    this.rootPos = [x, y, this.rootPos[2]];
    this.rootYaw = yawRad; this.rootRot = yaw(yawRad);
  }

  get cur() { return this.animFrame; }
  clipName() { return this.clipNames[this.animClip]; }
  // True only while riding the jump skill (run-up -> flight -> landing).
  get jumping() { return this.jumpLocked > 0; }

  // -- jump bucket (obstacle-triggered, mm_g1-style run-up + lock) --
  // Nearest forward distance (m) to an obstacle whose height band a standing body
  // overlaps, along the current facing; Infinity if none ahead. Arms the auto-jump.
  _forwardObstacleDist() {
    const fdir = [Math.cos(this.rootYaw), Math.sin(this.rootYaw)];
    const perp = [-fdir[1], fdir[0]];
    let best = Infinity;
    for (const o of this.env.obs) {
      if (o.h[0] > 1.3 || o.h[1] < 0.0) continue;       // not in standing-body band
      const relx = o.center[0] - this.rootPos[0], rely = o.center[1] - this.rootPos[1];
      const fwd = relx * fdir[0] + rely * fdir[1];
      if (fwd <= 0.0) continue;                          // behind us
      const reach = (o.isEllipse ? Math.max(o.ext[0], o.ext[1]) : o.radius) + 0.5;
      if (Math.abs(relx * perp[0] + rely * perp[1]) > reach) continue;   // off to the side
      if (fwd < best) best = fwd;
    }
    return best;
  }

  // Run-up frame whose features best continue the current pose (smooth take-off).
  _bestJumpEntry() {
    if (this.jumpEnter.length === 0) return null;
    const cb = this.animFrame * 27;
    let best = null, bd = Infinity;
    for (const f of this.jumpEnter) {
      const fb = f * 27; let d = 0;
      for (let i = 0; i < 27; i++) { const x = this.Xn[fb + i] - this.Xn[cb + i]; d += x * x; }
      if (d < bd) { bd = d; best = f; }
    }
    return best;
  }

  // Inertialize the pose discontinuity from the current frame into frame b, move there.
  _switchTo(b) {
    const a = this.animFrame;
    const da = this._row(this.dof, 29, a), dbb = this._row(this.dof, 29, b);
    const va = this._row(this.dofVel, 29, a), vb = this._row(this.dofVel, 29, b);
    for (let i = 0; i < 29; i++) { this.offDof[i] += da[i] - dbb[i]; this.offDofVel[i] += va[i] - vb[i]; }
    this.offH += this.rootPosDB[a * 3 + 2] - this.rootPosDB[b * 3 + 2];
    this.animFrame = b; this.animClip = this._clipOf(b);
  }

  _query(qhCtrl) {
    const q = new Array(27), Xs = this.Xstd, Xm = this.Xmean, b = this.animFrame * 27;
    // trajPos (0:6), trajDir (6:12)
    for (let k = 0; k < 3; k++) {
      const dp = quat.invMulVec(qhCtrl, v3.sub(this.Tpos[k], this.rootPos));
      q[2 * k] = dp[0]; q[2 * k + 1] = dp[1];
      const dd = quat.invMulVec(qhCtrl, this.Tdir[k]);
      q[6 + 2 * k] = dd[0]; q[6 + 2 * k + 1] = dd[1];
    }
    // pose (12:27) de-normalized from Xn then re-normalized below
    for (let i = 12; i < 27; i++) q[i] = this.Xn[b + i] * Xs[i] + Xm[i];
    for (let i = 0; i < 27; i++) q[i] = (q[i] - Xm[i]) / Xs[i];
    return q;
  }

  step(leftStick, rightStick) {
    const DT = this.DT;
    const desiredVel = v3.scale(leftStick, this.MAX_SPEED);
    this.targetSpeed = v3.norm(desiredVel);
    if (v3.norm(rightStick) > 0.01) this.desiredDir = v3.scale(rightStick, 1 / v3.norm(rightStick));
    else if (v3.norm(leftStick) > 0.01) this.desiredDir = v3.scale(leftStick, 1 / v3.norm(leftStick));
    const desiredRot = yaw(Math.atan2(this.desiredDir[1], this.desiredDir[0]));

    // desired trajectory springs (per horizon)
    this.Tpos = []; this.Tdir = [];
    for (let k = 0; k < 3; k++) {
      const [P] = trajectorySpringPosition(this.rootPos, this.rootVel, this.rootAcc, desiredVel, this.VELH, this.Ttimes[k]);
      this.Tpos.push(P);
      const [Q] = trajectorySpringRotation(this.rootRot, this.rootAng, desiredRot, this.ROTH, this.Ttimes[k]);
      this.Tdir.push(quat.mulVec(Q, FORWARD));
    }

    // nearby obstacles per tap (in heading-local frame)
    const tapWorld = this.Tpos.map((p) => [p[0], p[1]]);
    const { circ, ell } = this.env.getNearby([this.rootPos[0], this.rootPos[1]], this.rootYaw, tapWorld, this.horizons);

    // Jump bucket trigger: arm a jump the instant a wall crosses triggerDist ahead,
    // entered from its best-matching run-up and ridden through flight + landing
    // (search locked), exactly like mm_g1's J key but auto-triggered by the obstacle.
    if (this.jumpLocked === 0) {
      const fd = this._forwardObstacleDist();
      if (this.prevFd > this.triggerDist && this.triggerDist >= fd) {
        const entry = this._bestJumpEntry();
        if (entry !== null) {
          this._switchTo(entry);
          this.jumpLocked = Math.max(1, this.endOf.get(entry) - entry);
          this.searchTimer = this.SEARCH_TIME;
        }
      }
      this.prevFd = fd;
    }

    // Search (rate limited, LOCOMOTION bucket only): env-aware nearest pose.
    if (this.jumpLocked === 0 && this.searchTimer <= 0) {
      const qhCtrl = yaw(this.rootYaw);
      const Xq = this._query(qhCtrl);
      const pw = this.penaltyWeight * this.anticipation * Math.max(0.5, this.targetSpeed);
      const best = searchEnv(this.db, Xq, this.w, pw, circ, ell, this.threshold, this.heightMode, this.locoMask);
      if (best !== this.animFrame) {
        const a = this.animFrame;
        const da = this._row(this.dof, 29, a), dbb = this._row(this.dof, 29, best);
        const va = this._row(this.dofVel, 29, a), vb = this._row(this.dofVel, 29, best);
        for (let i = 0; i < 29; i++) { this.offDof[i] += da[i] - dbb[i]; this.offDofVel[i] += va[i] - vb[i]; }
        this.offH += this.rootPosDB[a * 3 + 2] - this.rootPosDB[best * 3 + 2];
        this.animFrame = best; this.animClip = this._clipOf(best);
      }
      this.searchTimer = this.SEARCH_TIME;
    }

    this._updateEvasion(circ, ell);

    // advance playhead
    this.animFrame = clamp(this.animFrame + 1, this.starts[this.animClip], this.stops[this.animClip] - 1);
    this.searchTimer -= DT;
    if (this.jumpLocked > 0) {
      this.jumpLocked -= 1;
      if (this.jumpLocked === 0) this.searchTimer = 0;     // search out of the jump at once
    } else if (this.animFrame >= this.stops[this.animClip] - 2) {
      this.searchTimer = 0;
    }
    const f = this.animFrame;

    // integrate root from matched clip root velocity
    const [, , acc] = trajectorySpringPosition(this.rootPos, this.rootVel, this.rootAcc, desiredVel, this.ROTH, DT);
    this.rootAcc = acc;
    const qhClip = yaw(this.thetaDB[f]);
    const clipVelLocal = quat.invMulVec(qhClip, this._row(this.rootVelDB, 3, f));
    this.rootVel = quat.mulVec(this.rootRot, clipVelLocal);
    this.rootAng = [0, 0, this.yawRateDB[f]];
    this.rootPos = v3.add(this.rootPos, v3.scale(this.rootVel, DT));
    this.rootYaw += this.yawRateDB[f] * DT;
    this.rootRot = yaw(this.rootYaw);

    // inertialized pose
    [this.offDof, this.offDofVel] = decaySpringDamper(this.offDof, this.offDofVel, this.INERT, DT);
    [this.offH, this.offHVel] = decaySpringDamperScalar(this.offH, this.offHVel, this.INERT, DT);
    const dofRow = this._row(this.dof, 29, f), dofOut = new Array(29);
    for (let i = 0; i < 29; i++) dofOut[i] = dofRow[i] + this.offDof[i];
    const tilt = quat.mul(quat.inv(qhClip), this._row(this.rootQuatDB, 4, f));
    const rootQuatOut = quat.mul(this.rootRot, tilt);

    const qpos = new Float64Array(36);
    qpos[0] = this.rootPos[0]; qpos[1] = this.rootPos[1];
    qpos[2] = this.rootPosDB[f * 3 + 2] + this.offH;
    qpos[3] = rootQuatOut[0]; qpos[4] = rootQuatOut[1]; qpos[5] = rootQuatOut[2]; qpos[6] = rootQuatOut[3];
    for (let i = 0; i < 29; i++) qpos[7 + i] = dofOut[i];
    return qpos;
  }

  // shrink the facing (trajDir) weight when an obstacle is close (port of _update_evasion)
  _updateEvasion(circ, ell) {
    const thr = this.threshold, f = this.animFrame, A = this.A;
    let closest = Infinity;
    for (let k = 0; k < 3; k++) {
      const center = [A.posTaps[f * 6 + k * 2], A.posTaps[f * 6 + k * 2 + 1]];
      const ax = [A.ellAxis[f * 6 + k * 2], A.ellAxis[f * 6 + k * 2 + 1]];
      const ext = [A.ellExtP[f * 3 + k], A.ellExtS[f * 3 + k]];
      for (const c of circ[k]) {
        let { dist } = eg.pointToEllipse(center, ax, ext, [c[0], c[1]]);
        dist = Math.max(dist - c[2], eg.MAX_INSIDE);
        if (dist < thr) closest = Math.min(closest, dist);
      }
      for (const e of ell[k]) {
        const d = eg.ellipseToEllipse(center, ax, ext, [e[0], e[1]], [e[2], e[3]], [e[4], e[5]]);
        if (d < thr) closest = Math.min(closest, d);
      }
    }
    let target = this.startDirW, rate;
    if (isFinite(closest) && closest < thr) {
      const factor = Math.log10(Math.max(closest / thr, 1e-3)) + 1.0;
      target = this.startDirW * Math.max(this.evasion, factor);
      rate = clamp(this.DT * 10.0, 0, 1);
    } else {
      rate = clamp(this.DT * 100.0, 0, 1);
    }
    const cur = this.w[6];
    const nw = cur + (target - cur) * rate;
    for (let i = 6; i < 12; i++) this.w[i] = nw;
  }
}
